pgx/src/memcxt.rs

```diff
@@ -181,12 +181,60 @@ pub struct OwnedMemoryContext {
 
 impl Drop for OwnedMemoryContext {
     fn drop(&mut self) {
+        extern "C" {
+            pub fn pgx_MemoryContextIsValid(context: pg_sys::MemoryContext) -> bool;
+        }
+
+        unsafe fn switch_context(
+            cxt: pg_sys::MemoryContext,
+            previous: pg_sys::MemoryContext,
+        ) -> bool {
+            if ptr::eq(pg_sys::CurrentMemoryContext, cxt) {
+                // If the previous context is no longer valid, switch to TopMemoryContext
+                pg_sys::CurrentMemoryContext = if pgx_MemoryContextIsValid(previous) {
+                    previous
+                } else {
+                    pg_sys::TopMemoryContext
+                };
+                true
+            } else {
+                false
+            }
+        }
+
+        unsafe fn switch_children(
+            cxt: pg_sys::MemoryContext,
+            previous: pg_sys::MemoryContext,
+        ) -> bool {
+            let mut child = (*cxt).firstchild;
+            while !child.is_null() {
+                if switch_context(child, previous) {
+                    return true;
+                }
+                if switch_children(child, previous) {
+                    return true;
+                }
+                child = (*child).nextchild;
+            }
+            return false;
+        }
```

Comment:
I find this "`bool`-returning mutation function" idiom to be confusing to reason about when used in a recursive situation. I prefer it when conditionals are "pure" in and of themselves.

Reply:
Very good point. I've simplified this code. Please let me know if this is better.